File: src/FluentCards/ActionSetBuilder.cs
namespace FluentCards;

/// <summary>
/// Fluent builder for creating ActionSet elements.
/// </summary>
public class ActionSetBuilder
{
    private readonly ActionSet _actionSet = new() { Actions = new List<AdaptiveAction>() };

    /// <summary>
    /// Sets the unique identifier for the action set.
    /// </summary>
    /// <param name="id">The unique identifier.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public ActionSetBuilder WithId(string id)
    {
        _actionSet.Id = id;
        return this;
    }

    /// <summary>
    /// Adds an action to the action set.
    /// </summary>
    /// <param name="configure">Action to configure the action.</param>
    /// <returns>The builder instance for method chaining.</returns>
    public ActionSetBuilder AddAction(Action<ActionBuilder> configure)
    {
        var builder = new ActionBuilder();
        configure(builder);
        _actionSet.Actions!.Add(builder.Build());
        return this;
    }

    /// <summary>
    /// Builds and returns the configured ActionSet.
    /// </summary>
    /// <returns>The configured ActionSet instance.</returns>
    public ActionSet Build()
    {
        return _actionSet;
    }
}

File: go/fluentcards/action_set_builder.go
package fluentcards

// ActionSetBuilder builds an ActionSet body element (a group of actions within the card body).
type ActionSetBuilder struct {
	data map[string]any
}

func newActionSetBuilder() *ActionSetBuilder {
	return &ActionSetBuilder{data: map[string]any{"type": "ActionSet", "actions": []any{}}}
}

func (b *ActionSetBuilder) WithID(id string) *ActionSetBuilder {
	b.data["id"] = id
	return b
}

func (b *ActionSetBuilder) WithSpacing(spacing Spacing) *ActionSetBuilder {
	b.data["spacing"] = string(spacing)
	return b
}

func (b *ActionSetBuilder) WithIsVisible(isVisible bool) *ActionSetBuilder {
	b.data["isVisible"] = isVisible
	return b
}

func (b *ActionSetBuilder) WithSeparator(separator bool) *ActionSetBuilder {
	b.data["separator"] = separator
	return b
}

func (b *ActionSetBuilder) WithHeight(height string) *ActionSetBuilder {
	b.data["height"] = height
	return b
}

func (b *ActionSetBuilder) WithFallback(fallback any) *ActionSetBuilder {
	b.data["fallback"] = fallback
	return b
}

func (b *ActionSetBuilder) WithRequires(key, version string) *ActionSetBuilder {
	reqs, ok := b.data["requires"].(map[string]any)
	if !ok {
		reqs = map[string]any{}
	}
	reqs[key] = version
	b.data["requires"] = reqs
	return b
}

func (b *ActionSetBuilder) WithRtl(rtl bool) *ActionSetBuilder {
	b.data["rtl"] = rtl
	return b
}

func (b *ActionSetBuilder) AddAction(configure func(*ActionBuilder)) *ActionSetBuilder {
	ab := newActionBuilder()
	configure(ab)
	actions := b.data["actions"].([]any)
	b.data["actions"] = append(actions, ab.Build())
	return b
}

func (b *ActionSetBuilder) Build() Card {
	return b.data
}
